flang/include/flang/Evaluate/tools.h

```diff
@@ -1240,6 +1240,24 @@ inline bool HasCUDAAttrs(const Expr<SomeType> &expr) {
   return false;
 }
 
+/// Check if the expression is a mix of host and device variables that require
+/// implicit data transfer.
+inline bool HasCUDAImplicitTransfer(const Expr<SomeType> &expr) {
+  unsigned hostSymbols = 0;
+  unsigned deviceSymbols = 0;
+  for (const Symbol &sym : CollectSymbols(expr)) {
+    if (const auto *details =
+            sym.GetUltimate().detailsIf<semantics::ObjectEntityDetails>()) {
```

Comment:
You may want to guard against derived type component symbol here, you may flag them as host while the base entity is on the device.

E.g:

```
  type t
    integer :: i(10)
  end type
  type(t), device :: tdev
  integer :: host(10)
  host = tdev%i + 10
```

`tdev%i + 10` would likely be flagged as requiring data transfer because `CollectSymbols` will collect both `tdev` and `i`, that I think both have `ObjectEntityDetails`, but where tdev would have the device attribute and not `i` (it is a symbol belonging to the type definition scope).

You can detect component symbols using `sym.owner().IsDerivedType()`.

I am assuming the following is not possible (even with pointer/allocatable attribute):

```
  type t
    integer, device :: i(10)
  end type
```

The current code is probably OK from a correctness point of view though (doing the transfer before the addition).

Reply:
Good point. I'll update that. 

> I am assuming the following is not possible (even with pointer/allocatable attribute):

So it turns out this is possible with allocatable component:
From 3.2.1
> Members of a derived type may not have the device attribute unless they are allocatable.

Reply:
> So it turns out this is possible with allocatable component

Then you should not apply my comment for allocatable components. BTW, what is the semantic of an allocatable component without the device attribute in a derived type instance that is device, is the allocatable component considered to be host or device?

```
 type t
  real, allocatable :: x(:)
 end type
 type(t), device :: a
 ! Is a%x considered to be on the host or on the device?
```

Reply:
The semantic is not clearly stated. If members of the derived-type are not allocatable or pointer they are on the device. I would say that this should also apply to allocatable. Talking with Brent it seems that the correct use is to make the derived type `managed` or `unified` when it has `allocatable` components. I'll work with him so that we can have this written in the specs.
I'll add a TODO until we have a more precise semantic on this.